File: kuhn.cpp
#include <vector>
#include <array>
#include <numeric>
#include <algorithm>
#include <map>
#include <iostream>
#include <random>
#include <chrono>

// Enum for card ranks
enum Card {
    JACK, QUEEN, KING
};

// Helper to get rank of a card for comparison
int get_rank_value(Card card) {
    if (card == JACK) return 0;
    if (card == QUEEN) return 1;
    if (card == KING) return 2;
    return -1; // Should not happen
}

std::string card_to_string(Card card) {
    switch (card) {
        case JACK: return "J";
        case QUEEN: return "Q";
        case KING: return "K";
    }
    return "Unknown";
}

// Enum for player actions
enum Action {
    FOLD, CHECK, CALL, BET
};

// Represents the state of a Kuhn Poker game
class KuhnState {
public:
    std::vector<Card> _player_cards; // Cards dealt to each player
    int _current_player;             // Current player to act (0 or 1)
    std::vector<int> _bets;          // Bets made by each player
    std::vector<Action> _history;    // History of actions taken in the current street

    KuhnState(std::vector<Card> player_cards, int current_player, std::vector<int> bets, std::vector<Action> history)
        : _player_cards(player_cards), _current_player(current_player), _bets(bets), _history(history) {}

    bool is_terminal() const {
        if (_history.empty()) return false;

        Action last_action = _history.back();
        if (last_action == FOLD) return true;

        if (_history.size() == 2) {
            // Cases: CHECK, CHECK or BET, CALL
            if (_history[0] == CHECK && _history[1] == CHECK) return true;
            if (_history[0] == BET && _history[1] == CALL) return true;
        }
        return false;
    }

    std::vector<Action> get_legal_actions() const {
        std::vector<Action> legal_actions;
        if (_history.empty()) {
            legal_actions.push_back(CHECK);
            legal_actions.push_back(BET);
        } else if (_history.size() == 1) {
            Action prev_action = _history[0];
            if (prev_action == CHECK) {
                legal_actions.push_back(CHECK);
                legal_actions.push_back(BET);
            } else if (prev_action == BET) {
                legal_actions.push_back(FOLD);
                legal_actions.push_back(CALL);
            }
        } else if (_history.size() == 2) {
            // This case handles when P0 bets, P1 bets, and it's P0's turn again
            Action prev_action = _history[1];
            if (prev_action == BET) {
                legal_actions.push_back(FOLD);
                legal_actions.push_back(CALL);
            }
        }
        return legal_actions;
    }

    KuhnState apply_action(Action action) const {
        std::vector<Action> new_history = _history;
        new_history.push_back(action);

        std::vector<int> new_bets = _bets;
        int new_current_player = 1 - _current_player;

        if (action == BET) {
            new_bets[_current_player] += 1; // Kuhn bet size is 1
        } else if (action == CALL) {
            // Player calls, matches the opponent's bet. The opponent's bet is _bets[1-_current_player] - _bets[_current_player]
            new_bets[_current_player] = new_bets[1 - _current_player];
        }
        return KuhnState(_player_cards, new_current_player, new_bets, new_history);
    }

    std::vector<double> get_returns() const {
        if (!is_terminal()) {
            return {}; // Should only be called on terminal states
        }

        std::vector<double> returns(_player_cards.size());

        if (_history.back() == FOLD) {
            int winner = _current_player; // The player who didn't fold
            int loser = 1 - _current_player;
            returns[winner] = _bets[loser]; // Winner gets loser's bet
            returns[loser] = -_bets[loser]; // Loser loses their bet
        } else {
            // Showdown: CHECK, CHECK or BET, CALL
            int winning_player = (_player_cards[0] > _player_cards[1]) ? 0 : 1;
            int losing_player = 1 - winning_player;
            
            double pot = _bets[0] + _bets[1];
            returns[winning_player] = pot - _bets[winning_player];
            returns[losing_player] = -_bets[losing_player];
        }
        return returns;
    }

    std::string to_string() const {
        std::string s = "Cards: [" + card_to_string(_player_cards[0]) + ", " + card_to_string(_player_cards[1]) + "] ";
        s += "Current Player: " + std::to_string(_current_player) + " ";
        s += "Bets: [" + std::to_string(_bets[0]) + ", " + std::to_string(_bets[1]) + "] ";
        s += "History: ";
        for (Action a : _history) {
            if (a == FOLD) s += "F ";
            else if (a == CHECK) s += "X ";
            else if (a == CALL) s += "C ";
            else if (a == BET) s += "B ";
        }
        return s;
    }
};

// Represents the Kuhn Poker game
class KuhnGame {
public:
    int _num_players;
    std::array<Card, 3> _deck;
    int _initial_pot_size;
    int _bet_size;
    int _raises_allowed; // Kuhn poker has at most 1 raise

    KuhnGame() : _num_players(2), _deck({JACK, QUEEN, KING}), _initial_pot_size(2), _bet_size(1), _raises_allowed(1) {}

    KuhnState get_initial_state() const {
        // Shuffle and deal cards
        std::vector<Card> shuffled_deck(_deck.begin(), _deck.end());
        unsigned seed = std::chrono::system_clock::now().time_since_epoch().count();
        std::default_random_engine rng(seed);
        std::shuffle(shuffled_deck.begin(), shuffled_deck.end(), rng);

        std::vector<Card> player_cards = {shuffled_deck[0], shuffled_deck[1]};
        std::vector<int> bets = {1, 1}; // Each player posts an ante of 1
        std::vector<Action> history;

        return KuhnState(player_cards, 0, bets, history); // Player 0 starts
    }

    int get_num_players() const {
        return _num_players;
    }
};

int main() {
    // Seed the random number generator (already done inside get_initial_state for shuffling)
    // unsigned seed = std::chrono::system_clock::now().time_since_epoch().count();
    // std::default_random_engine rng(seed);

    KuhnGame game;
    KuhnState state = game.get_initial_state();

    std::cout << "Initial State: " << state.to_string() << std::endl;

    // Example game play (simplified)
    while (!state.is_terminal()) {
        std::vector<Action> legal_actions = state.get_legal_actions();
        if (legal_actions.empty()) break;

        // For demonstration, just pick the first legal action
        Action chosen_action = legal_actions[0];
        std::cout << "Player " << state._current_player << " chooses action: ";
        if (chosen_action == FOLD) std::cout << "FOLD";
        else if (chosen_action == CHECK) std::cout << "CHECK";
        else if (chosen_action == CALL) std::cout << "CALL";
        else if (chosen_action == BET) std::cout << "BET";
        std::cout << std::endl;

        state = state.apply_action(chosen_action);
        std::cout << "New State: " << state.to_string() << std::endl;
    }

    if (state.is_terminal()) {
        std::vector<double> returns = state.get_returns();
        std::cout << "Terminal State. Returns: [" << returns[0] << ", " << returns[1] << "]" << std::endl;
    }

    return 0;
}

File: kuhn.py
import random
from enum import Enum

# Enums for card ranks and player actions
class Card(Enum):
    JACK = 0
    QUEEN = 1
    KING = 2

class Action(Enum):
    FOLD = 0
    CHECK = 1
    CALL = 2
    BET = 3

def card_to_string(card: Card) -> str:
    if card == Card.JACK: return "J"
    if card == Card.QUEEN: return "Q"
    if card == Card.KING: return "K"
    return "Unknown"


class KuhnState:
    def __init__(self, player_cards: list[Card], current_player: int, bets: list[int], history: list[Action]):
        self._player_cards = player_cards
        self._current_player = current_player
        self._bets = bets
        self._history = history

    def is_terminal(self) -> bool:
        if not self._history: # Equivalent to .empty() check
            return False

        last_action = self._history[-1]
        if last_action == Action.FOLD:
            return True

        # Terminal cases for Kuhn poker:
        # 1. CHECK, CHECK (both players check)
        # 2. BET, CALL (one player bets, other calls)
        # 3. CHECK, BET, CALL (check, bet, call sequence)
        # 4. BET, FOLD (bet and fold, handled above)
        # 5. CHECK, BET, FOLD (check, bet, fold, handled above)
        
        if len(self._history) == 2:
            # Cases: CHECK, CHECK or BET, CALL
            if self._history[0] == Action.CHECK and self._history[1] == Action.CHECK:
                return True
            if self._history[0] == Action.BET and self._history[1] == Action.CALL:
                return True
        elif len(self._history) == 3:
            # Case: CHECK, BET, CALL
            if (self._history[0] == Action.CHECK and 
                self._history[1] == Action.BET and 
                self._history[2] == Action.CALL):
                return True
        return False

    def get_legal_actions(self) -> list[Action]:
        legal_actions = []
        if not self._history:
            legal_actions.append(Action.CHECK)
            legal_actions.append(Action.BET)
        elif len(self._history) == 1:
            prev_action = self._history[0]
            if prev_action == Action.CHECK:
                legal_actions.append(Action.CHECK)
                legal_actions.append(Action.BET)
            elif prev_action == Action.BET:
                legal_actions.append(Action.FOLD)
                legal_actions.append(Action.CALL)
        elif len(self._history) == 2:
            # This case handles when P0 bets, P1 bets, and it's P0's turn again
            prev_action = self._history[1]
            if prev_action == Action.BET:
                legal_actions.append(Action.FOLD)
                legal_actions.append(Action.CALL)
        return legal_actions

    def apply_action(self, action: Action):
        new_history = list(self._history) # Create a copy
        new_history.append(action)

        new_bets = list(self._bets) # Create a copy
        new_current_player = 1 - self._current_player

        if action == Action.BET:
            new_bets[self._current_player] += 1 # Kuhn bet size is 1
        elif action == Action.CALL:
            # Player calls, matches the opponent's bet.
            # The opponent's bet is new_bets[1 - self._current_player] - new_bets[self._current_player]
            new_bets[self._current_player] = new_bets[1 - self._current_player]
        return KuhnState(self._player_cards, new_current_player, new_bets, new_history)

    def get_returns(self) -> list[float]:
        if not self.is_terminal():
            return [] # Should only be called on terminal states

        returns = [0.0, 0.0] # Assuming 2 players based on the C++ code

        if self._history[-1] == Action.FOLD:
            winner = self._current_player # The player who didn't fold
            loser = 1 - self._current_player
            returns[winner] = float(self._bets[loser]) # Winner gets loser's bet
            returns[loser] = -float(self._bets[loser]) # Loser loses their bet
        else:
            # Showdown: CHECK, CHECK or BET, CALL
            winning_player = 0 if self._player_cards[0].value > self._player_cards[1].value else 1
            losing_player = 1 - winning_player
            
            pot = float(self._bets[0] + self._bets[1])
            returns[winning_player] = pot - self._bets[winning_player]
            returns[losing_player] = -float(self._bets[losing_player])
        return returns

    def __str__(self) -> str:
        s = f"Cards: [{card_to_string(self._player_cards[0])}, {card_to_string(self._player_cards[1])}] "
        s += f"Current Player: {self._current_player} "
        s += f"Bets: [{self._bets[0]}, {self._bets[1]}] "
        s += "History: "
        for a in self._history:
            if a == Action.FOLD: s += "F "
            elif a == Action.CHECK: s += "X "
            elif a == Action.CALL: s += "C "
            elif a == Action.BET: s += "B "
        return s.strip()


class KuhnGame:
    def __init__(self):
        self._num_players = 2
        self._deck = [Card.JACK, Card.QUEEN, Card.KING]
        self._initial_pot_size = 2
        self._bet_size = 1
        self._raises_allowed = 1 # Kuhn poker has at most 1 raise

    def get_initial_state(self):
        # Shuffle and deal cards
        shuffled_deck = list(self._deck) # Create a mutable copy
        random.shuffle(shuffled_deck)

        player_cards = [shuffled_deck[0], shuffled_deck[1]]
        bets = [1, 1] # Each player posts an ante of 1
        history = []

        return KuhnState(player_cards, 0, bets, history) # Player 0 starts

    def get_num_players(self) -> int:
        return self._num_players


if __name__ == "__main__":
    game = KuhnGame()
    state = game.get_initial_state()

    print(f"Initial State: {state}")

    # Example game play (simplified)
    while not state.is_terminal():
        legal_actions = state.get_legal_actions()
        if not legal_actions: # Check if empty
            break

        # For demonstration, just pick the first legal action
        chosen_action = legal_actions[0]
        print(f"Player {state._current_player} chooses action: ", end="")
        if chosen_action == Action.FOLD: print("FOLD")
        elif chosen_action == Action.CHECK: print("CHECK")
        elif chosen_action == Action.CALL: print("CALL")
        elif chosen_action == Action.BET: print("BET")

        state = state.apply_action(chosen_action)
        print(f"New State: {state}")

    if state.is_terminal():
        returns = state.get_returns()
        print(f"Terminal State. Returns: [{returns[0]}, {returns[1]}]") 